### PR TITLE
Use time of create MPU on backend storage as time of MPU

### LINKED ISSUE #8303: BUG: Multipart upload: mtime difference between storage and lakeFS can be substantial
For example: In S3 the mtime is determined when creating the multipart upload requests, while lakeFS mtime is determined upon completion of multipart upload.
Needless to say this can result in a very huge diff between the S3 mtime and lakeFS mtime.

Need to find a generic solution to this which will be valid for all storage adapters
Possible solution:
Upon CompleteMultipartUpload, stat the object on the blockstore and use the mtime to create the lakeFS entry.

In order to properly test this - we need to consider adding a head object interface to our block adapter.

### PR DESCRIPTION
**Pull request by arielshaqed**
This time is either reported by the block adapter during completion, or it can be read by stat'ting the underlying object after complete-MPU.

Fixes #8303.

### CODE REVIEW

**Comment by arielshaqed** [comment]
#### _Not_ a customer issue

This does _not_ match any report by a customer of an inconsistency.  It's just a relevant cleanup.

File: pkg/api/controller.go [hunk 1]
```diff
@@ -384,6 +384,12 @@ func (c *Controller) CompletePresignMultipartUpload(w http.ResponseWriter, r *ht
 
 	writeTime := time.Now()
 	checksum := httputil.StripQuotesAndSpaces(mpuResp.ETag)
+	if mpuResp.MTime != nil {
+		// Anything else can be _really_ wrong when the storage layer assigns the time
+		// of MPU creation.  For instance, the S3 block adapter takes sure to return an
```
Comment (N-o-Z) [inline pkg/api/controller.go:389]:
```suggestion
		// of MPU creation.  For instance, the S3 block adapter makes sure to return an
```

File: pkg/api/controller.go [hunk 2]
```diff
@@ -384,6 +384,12 @@ func (c *Controller) CompletePresignMultipartUpload(w http.ResponseWriter, r *ht
 
 	writeTime := time.Now()
 	checksum := httputil.StripQuotesAndSpaces(mpuResp.ETag)
+	if mpuResp.MTime != nil {
```
Comment (N-o-Z) [inline pkg/api/controller.go:387]: Can we put line 387 right above this line

File: pkg/gateway/operations/putobject.go [hunk 3]
```diff
@@ -309,7 +309,8 @@ func handlePut(w http.ResponseWriter, req *http.Request, o *PathOperation) {
 	// write metadata
 	metadata := amzMetaAsMetadata(req)
 	contentType := req.Header.Get("Content-Type")
-	err = o.finishUpload(req, blob.Checksum, blob.PhysicalAddress, blob.Size, true, metadata, contentType)
+	// BUG(ariels): Read MTime from upload!
```
Comment (N-o-Z) [inline pkg/gateway/operations/putobject.go:312]: I think we still haven't decided if this should be fixed or not - anyway it's not related to the changes in this PR (let's track it in an issue instead?)

File: pkg/block/s3/adapter.go [hunk 4]
```diff
@@ -779,6 +779,7 @@ func (a *Adapter) CompleteMultiPartUpload(ctx context.Context, obj block.ObjectP
 	etag := strings.Trim(aws.ToString(resp.ETag), `"`)
 	return &block.CompleteMultiPartUploadResponse{
 		ETag:             etag,
+		MTime:            headResp.LastModified,
```
Comment (N-o-Z) [inline pkg/block/s3/adapter.go:782]: 👍🏽

File: esti/multipart_test.go [hunk 5]
```diff
@@ -44,6 +52,11 @@ func TestMultipartUpload(t *testing.T) {
 	}
 
 	completedParts := uploadMultipartParts(t, ctx, logger, resp, parts, 0)
+
+	// On S3 the object should have Last-Modified time at around time of MPU creation.
+	// Ensure lakeFS fails the test if it fakes it by using the current time.
+	time.Sleep(2 * timeResolution)
```
Comment (N-o-Z) [inline esti/multipart_test.go:58]: Never too happy about hardcoded sleeps in tests but seems inevitable.
Perhaps scope it to blockstoreType == s3?

Comment (arielshaqed) [reply esti/multipart_test.go]: I completely understand.  Don't like it but did it, thanks!

File: pkg/api/controller.go (hunk 1 above)
Comment (arielshaqed) [reply pkg/api/controller.go]: Thanks

File: pkg/block/s3/adapter.go (hunk 4 above)
Comment (arielshaqed) [reply pkg/block/s3/adapter.go:782]: :-)

File: pkg/gateway/operations/putobject.go (hunk 3 above)
Comment (arielshaqed) [reply pkg/gateway/operations/putobject.go]: Removed.

**Review by arielshaqed — COMMENTED** [review]
Thanks!

Pulling once Esti passes again...

**Pull request merged**